src/y0/algorithm/falsification.py

```diff
@@ -0,0 +1,156 @@
+import copy
+from collections import abc
+from itertools import chain, combinations
+
+import networkx as nx
+import pandas as pd
+from ananke.graphs import SG
+from tqdm import tqdm
+
+from ..util.stat_utils import cressie_read
+
+
+class Result:
+    """By default, acts like a boolean, but also caries evidence graph."""
+
+    def __init__(self, separated: bool, a, b, given, evidence: nx.Graph):
+        """separated -- T/F judgement
+           a/b/given -- The question asked
+           evidence -- The end graph
+        """
+        self.separated = separated
+        self.a = a
+        self.b = b
+        self.given = given
+        self.evidence = evidence
+
+    def __bool__(self):
+        return self.separated
+
+    def __repr__(self):
+        return f"{repr(self.separated)} ('{self.a}' d-sep '{self.b}' given {self.given})"
+
+    def __eq__(self, other):
+        return self.separated == other
+
+
+def are_d_separated(graph: SG, a, b, *, given=frozenset()) -> Result:
+    """Tests if nodes named by a & b are d-separated in G.
+
+    Given conditions can be provided with the optional 'given' parameter.
+    returns T/F and the final graph (as evidence)
+    """
+    named = {a, b}.union(given)
+
+    # Filter to ancestors
+    keep = graph.ancestors(named)
+    graph = copy.deepcopy(graph.subgraph(keep))
+
+    # Moralize (link parents of mentioned nodes)
+    for u, v in get_augments(graph):
+        graph.add_udedge(u, v)
+
+    # disorient & remove givens
+    evidence_graph = disorient(graph)
+
+    keep = set(evidence_graph.nodes) - set(given)
+    evidence_graph = evidence_graph.subgraph(keep)
+
+    # check for path....
+    separated = not nx.has_path(evidence_graph, a, b)  # If no path, then d-separated!
+
+    return Result(separated, a, b, given=given, evidence=evidence_graph)
+
+
+def disorient(graph: SG) -> nx.Graph:
+    """Disorient the ananke segregated graph to a simple networkx graph."""
+    rv = nx.Graph()
+    rv.add_nodes_from(graph.vertices)
+    rv.add_edges_from(chain(graph.di_edges, graph.ud_edges, graph.bi_edges))
+    return rv
+
+
+def get_augments(graph: SG):
+    parents = [graph.parents([v]) for v in graph.vertices]
+    augments = [*chain(*[combinations(nodes, 2) for nodes in parents if len(parents) > 1])]
```

Comment:
In general rather than doing `[*expr]` you can just use the `list(expr)` function. Same can be said that `chain(*expr)` is the same as `chain.from_iterable(expr)`

```suggestion
    auguments = list(chain.from_iterable(combinations(nodes, 2) for nodes in parents if len(parents) > 1]))
```

Reply:
I generally prefer the `[*...]` notation to the `list(...)` notation.  Is there reason why you would discourage the symbolic notation?

`chain.from_iterable` makes sense.

Reply:
Not all readers will understand `[*...]` and it's uncommonly used, which further makes reading the code harder to understand as a whole

Reply:
I hope you don't mind I was doing a bit of maintenance and rewriting parts as I was trying to figure them out